File: fixtures/conflicting-status-sources/sdk-bridge.ts
// Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
//
// Runs this family's vectors against the npm reference SDK, agent-passport-system,
// and records per claim whether the SDK has an API for the behaviour the claim
// needs. Where it does not, the claim is recorded not_supported with the reason,
// and nothing is simulated in its place.
//
// The family's model is protocol neutral. This bridge is the one place where it
// touches APS, and it touches it in two ways only:
//
//   claim 1  project the reference decision onto the single revocation answer
//            verifyAuthorityDelegationChain accepts, and check the chain verdict,
//   claim 2  re-derive each source line's within_bound using the SDK freshness
//            primitives rather than the harness arithmetic.
//
// Claims 3 to 6 are recorded, not run, because the SDK has no API for them.
//
// From the suite root:
//
//     npm run verify:conflicting-status-sources:sdk-ts

import { readFileSync } from 'node:fs'
import { createHash } from 'node:crypto'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import {
  buildRevocationObservation,
  createLocalSigner,
  createSnapshotFreshness,
  enforceFreshnessPolicy,
  verifyAuthorityDelegationChain,
  verifyRevocationObservation,
} from 'agent-passport-system'

const here = dirname(fileURLToPath(import.meta.url))
const pkg = JSON.parse(
  readFileSync(join(here, '..', '..', 'node_modules', 'agent-passport-system', 'package.json'), 'utf8'),
) as { name: string; version: string }

const vectors = JSON.parse(readFileSync(join(here, 'vectors.json'), 'utf8')) as any
const chainFixture = JSON.parse(readFileSync(join(here, 'chain.json'), 'utf8')) as {
  now: string
  verification_keys: Record<string, string>
  chain: any[]
}

console.log(`implementation: ${pkg.name} ${pkg.version} (npm)`)
console.log(`node: ${process.version}`)
console.log('')

const failures: string[] = []

// ── claim 1: chain verdict under the projected single answer ──────────────
//
// The SDK's resolveRevocation returns one answer for one delegation. A
// conflict has no representation in that vocabulary, so it projects to
// 'unknown', which is the SDK's indeterminate input. The projection is the
// claim being tested: an implementation whose only status vocabulary is
// active/revoked/unknown must not answer valid for anything this family
// denies or leaves unestablished.
type Projection = { answer: string; expect_state: string; expect_code: string | null }

function project(decision: string, reason: string): Projection {
  if (decision === 'admit') return { answer: 'active', expect_state: 'valid', expect_code: null }
  if (decision === 'deny' && reason === 'status_revoked') {
    return { answer: 'revoked', expect_state: 'invalid', expect_code: 'REVOKED' }
  }
  // conflict and every not_established reason: the SDK has no value for either,
  // so both project to unknown.
  return { answer: 'unknown', expect_state: 'indeterminate', expect_code: 'REVOCATION_UNKNOWN' }
}

console.log('claim 1  chain verdict under the projected single revocation answer: SUPPORTED')
console.log('         api: verifyAuthorityDelegationChain({ resolveRevocation })')
let claim1 = 0
let claim1Total = 0
for (const c of vectors.cases) {
  for (const b of c.boundaries) {
    claim1Total += 1
    const p = project(b.expected.decision, b.expected.reason)
    const result: any = verifyAuthorityDelegationChain(chainFixture.chain, {
      now: chainFixture.now,
      resolveVerificationKey: (_issuer: string, method: string) =>
        chainFixture.verification_keys[method] ?? null,
      trustRoot: () => true,
      resolveRevocation: (() => p.answer) as any,
    })
    const code = Array.isArray(result.failures) && result.failures[0] ? result.failures[0].code : null
    const ok = result.state === p.expect_state && code === p.expect_code
    const neverValid = b.expected.decision === 'admit' || result.state !== 'valid'
    if (ok && neverValid) claim1 += 1
    else {
      failures.push(
        `claim 1 ${b.boundary_id}: projected ${p.answer}, wanted ${p.expect_state}/${p.expect_code}, ` +
        `got ${result.state}/${code}`,
      )
    }
    console.log(
      `         ${b.boundary_id.padEnd(12)} ${b.expected.decision.padEnd(16)}` +
      `-> resolveRevocation '${p.answer}' -> ${result.state}` +
      (code ? `/${code}` : ''),
    )
  }
}
console.log(`         ${claim1}/${claim1Total} boundaries matched`)
console.log('')

// ── claim 2: per-source freshness boundary ────────────────────────────────
console.log('claim 2  per-source freshness bound, answer inside or past it: SUPPORTED')
console.log('         api: enforceFreshnessPolicy, mode bounded_staleness')
console.log(
  '         note: recordRevocationFreshness, which builds the underlying record, is not\n' +
  '         exported from the package root; the composition is.',
)
let claim2 = 0
let claim2Total = 0
for (const c of vectors.cases) {
  for (const b of c.boundaries) {
    for (const line of b.expected.record.sources_consulted) {
      if (line.as_of === null) continue // an unavailable answer dates nothing
      claim2Total += 1
      const boundMs = line.freshness_bound_s * 1000
      const decision = enforceFreshnessPolicy(
        {
          source: line.source_id,
          maxStalenessMs: boundMs,
          checkedAt: new Date(b.evaluated_at),
          freshness: createSnapshotFreshness(line.as_of, line.freshness_bound_s),
        },
        { mode: 'bounded_staleness', boundedStalenessMs: boundMs, action_on_stale: 'deny' },
      )
      const sdkWithinBound = decision.effect === 'allow'
      if (sdkWithinBound === line.within_bound) claim2 += 1
      else {
        failures.push(
          `claim 2 ${b.boundary_id}/${line.source_id}: harness within_bound=${line.within_bound}, ` +
          `SDK effect=${decision.effect} (${decision.reason})`,
        )
      }
    }
  }
}
console.log(`         ${claim2}/${claim2Total} source lines matched the harness freshness boundary`)
console.log('')

// ── claim 3: two sources compared for conflict ────────────────────────────
console.log('claim 3  two status sources compared for conflict about one delegation: NOT_SUPPORTED')
console.log(
  '         reason: a RevocationObservation carries exactly one status_source, and\n' +
  '         decideFreshness takes exactly one RevocationFreshnessRecord. No exported\n' +
  '         function takes two answers about one authority_ref and returns a conflict.\n' +
  '         Nothing was simulated in its place.',
)
console.log('')

// ── claim 4: coverage over a declared source set ──────────────────────────
console.log('claim 4  coverage over a declared required-source set: NOT_SUPPORTED')
console.log(
  '         reason: no exported API takes a list of sources a verifier requires, so\n' +
  '         "every required source answered" is not expressible. The suite\'s\n' +
  '         status_coverage_incomplete reason has no SDK counterpart.',
)
console.log('')

// ── claim 5: a third verdict at the observation layer ─────────────────────
console.log('claim 5  not_established as an observation-layer outcome: NOT_SUPPORTED')
console.log(
  '         reason: FreshnessDecision.effect is allow | deny, and the SDK types say of\n' +
  '         RevocationObservationDecision that "No third verdict value exists". The\n' +
  '         chain layer does carry indeterminate, which claim 1 exercises; the two\n' +
  '         layers do not share a vocabulary.',
)
console.log('')

// ── claim 6: what an offline admission records ────────────────────────────
console.log('claim 6  an offline admission records the snapshot and the age it used: PARTIAL')
const offline = vectors.cases
  .flatMap((c: any) => c.boundaries)
  .find((b: any) => b.expected.record.verifier_mode === 'offline' && b.expected.decision === 'admit')
if (!offline) {
  failures.push('claim 6: no offline admit boundary in vectors.json')
} else {
  const snap = offline.expected.record.snapshot
  const boundMs = snap.declared_bound_s * 1000
  const decision = enforceFreshnessPolicy(
    {
      source: snap.source_id,
      maxStalenessMs: boundMs,
      checkedAt: new Date(offline.evaluated_at),
      freshness: createSnapshotFreshness(snap.as_of, snap.declared_bound_s),
    },
    { mode: 'bounded_staleness', boundedStalenessMs: boundMs },
  )
  const record = decision.record
  const signerKey = createHash('sha256')
    .update(`${vectors.determinism.seed_input}:sdk-bridge-observer:v1`, 'utf8')
    .digest('hex')
  const signed = await buildRevocationObservation(
    {
      authority_ref: vectors.delegation_ref,
      status_source: { kind: 'source', id: snap.source_id },
      observed_at: new Date(offline.evaluated_at),
      maximum_staleness_ms: boundMs,
      decision: { effect: decision.effect === 'allow' ? 'allow' : 'deny' },
    },
    createLocalSigner({ privateKeyHex: signerKey }),
  )
  const verified = verifyRevocationObservation(signed)
  console.log(`         boundary: ${offline.boundary_id}`)
  console.log(`         decideFreshness effect=${decision.effect}`)
  console.log(`         decideFreshness reason="${decision.reason}"`)
  console.log(
    `         RevocationFreshnessRecord carries: source=${record.source}, ` +
    `freshness.validAt=${record.freshness?.validAt}, maxStalenessMs=${record.maxStalenessMs}`,
  )
  console.log(
    `         SignedRevocationObservation fields: ${Object.keys(signed).sort().join(', ')}`,
  )
  console.log(`         verifyRevocationObservation: valid=${verified.valid}`)
  console.log(
    '         supported: the source identity and the tolerated staleness are structured\n' +
    '         fields on the signed observation, and the snapshot\'s own as_of is a\n' +
    '         structured field on the unsigned freshness record.\n' +
    '         not supported: the signed observation carries neither the snapshot as_of\n' +
    '         nor the age it admitted on, so a reader of the signed record alone cannot\n' +
    '         recompute either. On this path the decision short-circuits on result\n' +
    '         \'fresh\' and its reason carries no age at all; an age appears in the\n' +
    '         reason string only on the bounded-staleness branch, which is reached only\n' +
    '         after the source\'s own maxAge has already expired, and even there it is\n' +
    '         prose rather than a field.',
  )
  if (!verified.valid) failures.push('claim 6: the SDK did not verify its own observation')
  if (decision.effect !== 'allow') {
    failures.push(`claim 6: SDK denied a snapshot the fixture admits (${decision.reason})`)
  }
}
console.log('')

if (failures.length > 0) {
  for (const f of failures) console.error(`FAIL ${f}`)
  console.error(`\nconflicting-status-sources SDK bridge (TypeScript): ${failures.length} failure(s)`)
  process.exit(1)
}
console.log(
  `PASSED: conflicting-status-sources SDK bridge (TypeScript), ${pkg.name} ${pkg.version}. ` +
  `2 claims supported and matched, 1 partial, 3 recorded not_supported.`,
)

File: fixtures/conflicting-status-sources/sdk_bridge.py
#!/usr/bin/env python3
# Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
"""Runs this family's vectors against the PyPI reference SDK, agent-passport-system.

Same six claims as sdk-bridge.ts. The two SDKs do not have the same surface
here, so the claim table differs, and nothing is simulated to make them match.

Run against the SDK under test, for example:

    /path/to/venv/bin/python fixtures/conflicting-status-sources/sdk_bridge.py

Exit 0 when every supported claim matched, 1 otherwise, 2 when the SDK is not
importable.
"""
from __future__ import annotations

import importlib
import importlib.metadata as md
import json
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent

try:
    from agent_passport.v2.authority_delegation import verify_authority_delegation_chain
except Exception as err:  # pragma: no cover - environment problem, not a result
    print(f"agent-passport-system is not importable: {err}", file=sys.stderr)
    sys.exit(2)

VERSION = md.version("agent-passport-system")

vectors = json.loads((HERE / "vectors.json").read_text(encoding="utf-8"))
chain_fixture = json.loads((HERE / "chain.json").read_text(encoding="utf-8"))

print(f"implementation: agent-passport-system {VERSION} (PyPI)")
print(f"python: {sys.version.split()[0]}")
print("")

failures: list[str] = []


def project(decision: str, reason: str):
    if decision == "admit":
        return "active", "valid", None
    if decision == "deny" and reason == "status_revoked":
        return "revoked", "invalid", "REVOKED"
    return "unknown", "indeterminate", "REVOCATION_UNKNOWN"


print("claim 1  chain verdict under the projected single revocation answer: SUPPORTED")
print("         api: verify_authority_delegation_chain(resolve_revocation=...)")
matched = 0
total = 0
for case in vectors["cases"]:
    for boundary in case["boundaries"]:
        total += 1
        answer, want_state, want_code = project(
            boundary["expected"]["decision"], boundary["expected"]["reason"]
        )
        result = verify_authority_delegation_chain(
            chain_fixture["chain"],
            now=chain_fixture["now"],
            resolve_verification_key=lambda _issuer, method, _issued_at:
                chain_fixture["verification_keys"].get(method),
            trust_root=lambda _root: True,
            resolve_revocation=lambda _delegation, _answer=answer: _answer,
        )
        code = result.failures[0].code if result.failures else None
        ok = result.state == want_state and code == want_code
        never_valid = boundary["expected"]["decision"] == "admit" or result.state != "valid"
        if ok and never_valid:
            matched += 1
        else:
            failures.append(
                f"claim 1 {boundary['boundary_id']}: projected {answer}, wanted "
                f"{want_state}/{want_code}, got {result.state}/{code}"
            )
        suffix = f"/{code}" if code else ""
        print(
            f"         {boundary['boundary_id']:<12} "
            f"{boundary['expected']['decision']:<16}-> resolve_revocation '{answer}' "
            f"-> {result.state}{suffix}"
        )
print(f"         {matched}/{total} boundaries matched")
print("")

# The remaining five claims. Absence is asserted against the installed package
# rather than asserted from memory, so a later release that adds the surface
# turns this run into a failure instead of a stale note.
ABSENT = [
    (
        "claim 2",
        "per-source freshness bound, answer inside or past it",
        ["decide_freshness", "enforce_freshness_policy", "create_snapshot_freshness",
         "compute_evidence_age", "is_evidence_fresh"],
        "this SDK exposes no freshness-policy or evidence-age surface, so a per-source\n"
        "         bound cannot be expressed. The TypeScript SDK at 7.1.0 does have one,\n"
        "         so the two reference implementations do not cover the same ground here.",
    ),
    (
        "claim 3",
        "two status sources compared for conflict about one delegation",
        ["RevocationObservation", "build_revocation_observation"],
        "resolve_revocation returns one answer for one delegation and no API takes two\n"
        "         answers about one delegation. Nothing was simulated in its place.",
    ),
    (
        "claim 4",
        "coverage over a declared required-source set",
        ["required_sources", "coverage"],
        "no API takes a list of sources a verifier requires.",
    ),
    (
        "claim 5",
        "not_established as an observation-layer outcome",
        ["FreshnessDecision", "RevocationObservationDecision"],
        "there is no observation layer in this SDK. The chain layer does return\n"
        "         indeterminate, which claim 1 exercises.",
    ),
    (
        "claim 6",
        "an offline admission records the snapshot and the age it used",
        ["build_revocation_observation", "verify_revocation_observation"],
        "no observation record exists to carry a snapshot identity, its as_of or its\n"
        "         age, so there is nothing to record into.",
    ),
]

root = importlib.import_module("agent_passport")
for label, title, symbols, reason in ABSENT:
    present = [s for s in symbols if hasattr(root, s)]
    if present:
        failures.append(
            f"{label}: recorded not_supported, but agent_passport now exports {present}. "
            f"Re-run the survey and rewrite the claim."
        )
    print(f"{label}  {title}: NOT_SUPPORTED")
    print(f"         checked absent: {', '.join(symbols)}")
    print(f"         reason: {reason}")
    print("")

if failures:
    for item in failures:
        print(f"FAIL {item}", file=sys.stderr)
    print(
        f"\nconflicting-status-sources SDK bridge (Python): {len(failures)} failure(s)",
        file=sys.stderr,
    )
    sys.exit(1)

print(
    f"PASSED: conflicting-status-sources SDK bridge (Python), agent-passport-system "
    f"{VERSION}. 1 claim supported and matched, 5 recorded not_supported."
)
